FEATURE/ElvaRivle/t5Dxp0AK

ElvaRivle opened the pull request:
FEATURE/ElvaRivle/t5Dxp0AK

--------------------------------------------------------------------
 MAIN
 * added everything for user registration (route, controller, service, repository)
 * added `password-policy.js` file for settings password rules
--------------------------------------------------------------------
 OTHER
 * performing auto-login upon successful user registration
 * ROLES NOT YET IMPLEMENTED !!!
--------------------------------------------------------------------

ElvaRivle committed:
FEATURE/ElvaRivle/t5Dxp0AK

--------------------------------------------------------------------
 MAIN
 * added route for user registration (route, controller, service, repository)
 * added `password-policy.js` file for settings password rules
--------------------------------------------------------------------
 OTHER
 * performing auto-login upon successful user registration
--------------------------------------------------------------------

ElvaRivle committed:
FEATURE/ElvaRivle/t5Dxp0AK

--------------------------------------------------------------------
 MAIN
 * added 2 roles: `User` and `Admin`
 * added validator for roles and password
 * added `role` field to User model
--------------------------------------------------------------------
 OTHER
 * default role, if not specified, is `User`
 * .gitignore for IDE/text editor config files
--------------------------------------------------------------------
 TODO
 * move password hashing to mongoose `pre` middleware
 * use `joi` to validate request data in middleware
 * maybe implement nicer exception handling, middleware instead of catching in controller?
--------------------------------------------------------------------

ElvaRivle commented:
FEATURE/ElvaRivle/t5Dxp0AK

--------------------------------------------------------------------
 MAIN
 * added 2 roles: `User` and `Admin`
 * added validator for roles and password
 * added `role` field to User model
--------------------------------------------------------------------
 OTHER
 * default role, if not specified, is `User`
 * .gitignore for IDE/text editor config files
--------------------------------------------------------------------
 TODO
 * move password hashing to mongoose `pre` middleware 
 * use `joi` to validate request data in middleware
 * maybe implement nicer exception handling, middleware instead of catching in controller?
--------------------------------------------------------------------

becarselman commented on configuration/user-roles.js:
```diff
@@ -0,0 +1,4 @@
+module.exports = {
```
You can merge this file and password policy file into one utils/constants file.

becarselman commented on validators/roles.js:
```diff
@@ -0,0 +1,15 @@
+const userRoles = require("../configuration/user-roles")
+
+function doesRoleExist(role) {
+    for (const key in userRoles) {
```
You can use "some" function for this. And also you can merge password and roles validatior in one file. utils/validators.js

becarselman commented on repositories/user.js:
```diff
@@ -10,6 +10,25 @@ async function getUserByEmail(email) {
 }
 
+async function createUser(email, hashedPassword, role) {
+    if (!email) {
```
I think you are already check this in service. And also we need to care this in service not in repository

becarselman commented on repositories/user.js:
```diff
@@ -10,6 +10,25 @@ async function getUserByEmail(email) {
     return await User.findOne({ email: email })
 }
 
+async function createUser(email, hashedPassword, role) {
```
Please send this parameters as one object and use it like that.

becarselman commented on services/register.js:
```diff
@@ -0,0 +1,42 @@
+const passwordValidator = require("../validators/password")
+
+
+async function registerUser(email, password, role) {
```
also we can take here 1 parameter "user". and that parameter will have {email:dlsakfldkasf, password:laksdfjlasfj, role:USER}. I think that is req.body. 
and then you can use it like user.email, user.password...

ElvaRivle committed:
FEATURE/ElvaRivle/t5Dxp0AK

--------------------------------------------------------------------
 MAIN
 * added `PORT` and `FRONTEND_URL` env variables
 * allowed CORS for `FRONTEND_URL`
 * send entire `req.body` to service, unpack object there, not in controller
 * removed unnecessary null/undefined checks in repositories and other functions, keep those checks only in service
--------------------------------------------------------------------
 OTHER
 * added `joi` and `cors` packages
 * disabled auto login user upon registration
 * created utils folder and merged validator and constants into single file
 * added `User` role to seeded users
--------------------------------------------------------------------
 TODO
 * use `joi` middleware to validate incoming data
--------------------------------------------------------------------

ElvaRivle commented:
FEATURE/ElvaRivle/t5Dxp0AK

--------------------------------------------------------------------
 MAIN
 * added `PORT` and `FRONTEND_URL` env variables
 * allowed CORS for `FRONTEND_URL`
 * send entire `req.body` to service, unpack object there, not in controller
 * removed unnecessary null/undefined checks in repositories and other functions, keep those checks only in service
--------------------------------------------------------------------
 OTHER
 * added `joi` and `cors` packages
 * disabled auto login user upon registration
 * created utils folder and merged validator and constants into single file
 * added `User` role to seeded users
--------------------------------------------------------------------
 TODO
 * use `joi` middleware to validate incoming data 
 * make roles case insensitive
--------------------------------------------------------------------

ElvaRivle committed:
FEATURE/ElvaRivle/t5Dxp0AK

--------------------------------------------------------------------
 MAIN
 * created joi validation schemas for request body
 * created register and login middlewares that use joi validation schemas
 * business logic validation kept in services, middleware only does non-business validation
--------------------------------------------------------------------
 OTHER
 * added `joi-password-complexity` package
 * updated password policy object to adhere to password complexity package (renamed object keys)
 * upper-case user roles
--------------------------------------------------------------------

ElvaRivle commented:
FEATURE/ElvaRivle/t5Dxp0AK

--------------------------------------------------------------------
 MAIN
 * created joi validation schemas for request body
 * separated validation rules from validation schemas
 * created register and login middlewares that use joi validation schemas
 * business logic validation kept in services, middleware only does non-business validation
--------------------------------------------------------------------
 OTHER
 * added `joi-password-complexity` package
 * updated password policy object to adhere to password complexity package (renamed object keys)
 * upper-case user roles
--------------------------------------------------------------------

TarikVelic commented:
Great job, LGTM :)

TarikVelic reviewed (approved):
Great job, LGTM :)
